File: 백준/Gold/2589. 보물섬/보물섬.py
# 보물섬
import sys
input = sys.stdin.readline
from collections import deque

def solution():
    N, M = map(int,input().split())
    graph = []

    for i in range(N):
        row = list(input().rstrip())
        graph.append(row)

    maxx = 0
    direc = [(1, 0), (-1, 0), (0, 1), (0, -1)]

    for i in range(N):
        for j in range(M):
            if graph[i][j] != "L":
                continue
            
            visited = [[False]*M for _ in range(N)]
            que = deque([(i, j, 0)])

            while que:
                r, c, cnt= que.popleft()

                if visited[r][c]:
                    continue
                else:
                    visited[r][c] = True
                    maxx = max(maxx, cnt)

                for dx, dy in direc:
                    nr, nc = r + dx, c + dy

                    if nr < 0 or nc < 0 or nr >= N or nc >= M:
                        continue
                    elif graph[nr][nc] == "W" or visited[nr][nc]:
                        continue
                    else:
                        que.append((nr, nc, cnt+1))
    return maxx

print(solution())
        

File: 백준/Gold/2589. 보물섬/보물섬.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        int max = 0;
        char[][] graph = new char[N][M];
        int[] dx = {1, -1, 0, 0};
        int[] dy = {0, 0, 1, -1};

        for (int i = 0; i < N; i++) {
            String row = br.readLine();
            for (int j = 0; j < M; j++) {
                graph[i][j] = row.charAt(j);
            }
        }

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (graph[i][j] == 'W') {
                    continue;
                }

                boolean[][] visited = new boolean[N][M];
                LinkedList<int[]> que = new LinkedList<>();
                que.add(new int[]{i, j, 0});

                while (!que.isEmpty()) {
                    int[] tmp = que.pollFirst();
                    int r = tmp[0]; int c = tmp[1]; int cnt = tmp[2];

                    if (visited[r][c]) {
                        continue;
                    } else {
                        visited[r][c] = true;
                        max = Math.max(max, cnt);
                    }

                    for (int k = 0; k < 4; k++) {
                        int nr = r + dx[k], nc = c + dy[k];

                        if (nr < 0 || nr >= N || nc < 0 || nc >= M) {
                            continue;
                        } else if (graph[nr][nc] == 'W' || visited[nr][nc]) {
                            continue;
                        } else {
                            que.add(new int[]{nr, nc, cnt+1});
                        }
                    }
                }
            }
        }
        System.out.println(max);
    }
}
